spellbook/models/sudoswap/ethereum/sudoswap_ethereum_events.sql

```diff
@@ -0,0 +1,256 @@
+    ),
+
+    tokens_ethereum_nft as (
+        SELECT 
+            *
+        FROM {{ ref('tokens_ethereum_nft') }}
+    ),
+
+    nft_ethereum_aggregators as (
+        SELECT 
+            *
+        FROM {{ ref('nft_ethereum_aggregators') }}
+    ),
+
+--logic CTEs
+    swaps_w_fees as (
+        SELECT 
+            *
+        FROM (
+            SELECT
+                call_tx_hash
+                , call_block_time
+                , call_block_number
+                , contract_address as pair_address
+                , call_trace_address
+                , ownerfee 
+                , protocolfee 
+                , trade_category
+                , nftcontractaddress
+                , asset_recip
+                , call_from
+                , row_number() OVER (partition by call_tx_hash, contract_address, call_trace_address order by fee_update_time desc, protocolfee_updatetime desc) as ordering
+            FROM (
+                SELECT 
+                    swaps.*
+                    , COALESCE(fu.newfee, pc.initialfee)/1e18 as ownerfee --most recent ownerfee, depends on bonding curve to implement it correctly. See explanation in fee table schema.
+                    , COALESCE(fu.evt_block_time, pc.block_time) as fee_update_time
+                    , pfu.newMultiplier/1e18 as protocolfee --most recent protocolfee, depends on bonding curve to implement it correctly. See explanation in fee table schema.
+                    , pfu.evt_block_time as protocolfee_updatetime
+                    , pc.nftcontractaddress
+                    , pc.asset_recip
+                FROM swaps
+                JOIN pairs_created pc ON pc.pair_address = contract_address --remember swaps from other NFT addresses won't appear!
+                LEFT JOIN owner_fee_update fu ON swaps.call_block_time >= fu.evt_block_time AND swaps.contract_address = fu.contract_address
+                LEFT JOIN protocol_fee_update pfu ON swaps.call_block_time >= pfu.evt_block_time
+            ) a
+        ) b
+        WHERE ordering = 1 --we want to keep the most recent ownerfee and protocol fee for each individual call (trade)
+    ),
+
+    swaps_w_traces as (
+        -- we traces to get NFT and ETH transfer data because sudoswap doesn't emit any data in events for swaps, so we have to piece it together manually based on trace_address.
+        SELECT
+            sb.call_block_time
+            , sb.call_block_number
+            , sb.trade_category
+            , CASE WHEN sb.trade_category = 'Buy' THEN SUM(value)/(1+sb.ownerfee+sb.protocolfee)
+                ELSE SUM(value)/(1-sb.ownerfee-sb.protocolfee)
+                END as base_price
+            , SUM(value) as trade_price_eth --should give total value of the trade (buy or sell)
+            , ARRAY_AGG(distinct CASE WHEN substring(input,1,10)='0x42842e0e' THEN bytea2numeric_v2(substring(input,139,64))::int ELSE '' END) 
```

Comment:
i wonder if the "ELSE '' END" is what is forcing a string within the array. maybe instead of '' it should be NULL?

Reply:
Ah yeah that's probably it. Lemme fix

Reply:
okay, the db says array(int) now -- want to confirm?